ibis/formats/pyarrow.py

```diff
@@ -341,3 +346,46 @@ def to_polars(self, schema: Schema) -> pl.DataFrame:
 
         df = pl.from_arrow(self.obj)
         return PolarsData.convert_table(df, schema)
+
+
+class PyArrowDatasetProxy(TableProxy):
+    __slots__ = ("obj", "__dict__")
+    obj: V
+
+    def __init__(self, obj: V):
+        self.obj = obj
+
+    def __len__(self):
+        return self.obj.count_rows()
+
+    # pyarrow datasets are hashable, so we override the hash from TableProxy
+    def __hash__(self):
+        return hash(self.obj)
+
+    @cached_property
+    def _cache(self):
+        return self.obj.to_table()
+
+    def to_frame(self) -> pd.DataFrame:
+        """Convert this input to a pandas DataFrame."""
+        return self._cache.to_pandas()
+
+    def to_pyarrow(self, schema: Schema) -> pa.Table:
+        """Convert this input to a PyArrow Table."""
+        return self._cache
+
+    def to_pyarrow_lazy(self, schema: Schema) -> ds.Dataset:
```

Comment:
Could we call this `to_pyarrow_dataset` instead? `to_pyarrow_lazy` feels less clear to me.

Reply:
yep